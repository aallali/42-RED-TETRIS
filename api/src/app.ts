
import * as express from "express";
import * as cors from "cors"
import { Socket } from "socket.io";
import Game from "./Game"

import ROOM from "./Room"
import PLAYER from "./Player"
import { ROOM_MODE } from "./types"


const log = console.log
interface IPlayer {
	id: string;
	nickname: string;
}
interface IRoom {
	id: string;
	name: string;
	players: []
	admin: IPlayer
}
class App {

	public app: express.Application;
	public http: any
	public io: Socket
	private port: number = 4242
	private gameIns: Game
	constructor() {
		this.app = express();
		this.app.use(
			cors({
				origin: "*",
				credentials: true,
			})
		);
		this.http = require("http").Server(this.app);
		this.io = require("socket.io")(this.http, {
			pingInterval: 60000,
			cors: {
				origin: "*",
			},
		});
		this.routes()
		this.initSocketEvents()
		this.gameIns = new Game(this.io)

	}
	/**
	 * 
	 */
	private routes() {
		this.app.get("/", (req: any, res: any) => {
			console.log(Array.from(this.gameIns.players))
			res.send(Array.from(this.gameIns.players))
		});
	}
	private initSocketEvents() {
	}
	public listen() {
		console.clear()
		this.http.listen(this.port, () =>
			console.log(`ðŸš€ Server is running on PORT ${this.port}`)
		);
	}
}

export default new App();
