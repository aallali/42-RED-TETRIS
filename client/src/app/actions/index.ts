import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
