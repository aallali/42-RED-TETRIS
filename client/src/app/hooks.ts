import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
