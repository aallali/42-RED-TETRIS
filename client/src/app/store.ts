import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
