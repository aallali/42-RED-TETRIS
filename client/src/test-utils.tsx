// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }