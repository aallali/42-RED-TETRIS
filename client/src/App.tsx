
/* eslint-disable react-hooks/exhaustive-deps */
import Home from './pages/home'
import Lobby from './pages/lobby'
import Error from './components/Error';
import {
	getPlayerNickname
} from './reducers/player.reducer';
import PlayBoard from "./pages/playboard"
import { useAppDispatch, useAppSelector } from "./app/hooks"
import { getError } from './reducers/error.reducer';
import { SET_ERROR } from './app/actions';
import { socket } from "./app/hooks"
import { getGameTitle } from './reducers/game.reducer';
import { store } from "./app/store";
import { useEffect } from 'react';


const HashParser = (window_hash: string) => {
	const regexp = /(?<roomname>[a-zA-Z0-9]{1,10})\[(?<username>[a-zA-Z0-9]{1,10})\]/;
	const result = window_hash.match(regexp);
	if (result)
		return result;
	return false;
};
// const namevalidtor = (u: string, r: string) => {
// 	if (u.length <= 10 && r.length <= 10)
// 		return true
// 	return false
// }
function App() {
	const dispatch = useAppDispatch()
	const playerIngame = useAppSelector(getGameTitle)
	const playerNickname = useAppSelector(getPlayerNickname)
	const error = useAppSelector(getError)
	const checkhash = () => {
		const playername = playerNickname
		socket.emit("PLAYER_LEFT")
		if (window.location.hash.substring(1)) {
			let checkHash = HashParser(window.location.hash.substring(1))
			if (checkHash && checkHash.groups) {
				const { roomname, username } = checkHash.groups
				if (playername && username !== playername) {
					dispatch(SET_ERROR({ title: "Error :", message: "there is an username already set to this browser, try to leave first to register new account" }))
				} else {
					dispatch(SET_ERROR({ title: "", message: "" }))
					const gameMode = store.getState().game.mode
					socket.emit('JOIN_ROOM', { room: roomname, playerName: username, mode: gameMode });
				}

			} else {
				dispatch(SET_ERROR({ title: "hash_query_error", message: "" }))
			}
		} else {
			dispatch(SET_ERROR({ title: "", message: "" }))
			// dispatch(LEAVE_GAME())
			// dispatch(RESET_STATES())
			socket.emit("PLAYER_LEFT")
		}
	}
	useEffect(() => {
		checkhash()
		window.removeEventListener("hashchange", () => { })
		window.addEventListener('hashchange', function () {
			checkhash()
		}, true);
	}, [])

	return (
		<div>
			{error.title ?
				(<Error title={error.title} message={error.message} />)
				: (!playerNickname ?
					(<Home />)
					: (playerIngame ?
						(<PlayBoard />)
						: (<Lobby />)))}

		</div>

	)
		;
}

export default App;
