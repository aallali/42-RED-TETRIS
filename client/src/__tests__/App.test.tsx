
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import App from '../App';
import * as reactRedux from "react-redux";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { aallaliStore, render, screen } from '../test-utils'
import { rest } from "msw";
import { setupServer } from "msw/node";
// const axios = require('axios');



const server = setupServer(
	// rest.get("http://localhost:4242/rooms", (_, res, ctx) => {
	// 	return res(ctx.status(200),ctx.json([]))
	// }),
	// rest.get("http://localhost:4242/top", (_, res, ctx) => {
	// 	return res(ctx.status(200), ctx.json({ data: `allali,67\nbot2,30\nbotUsername3,27` }))
	// })
);



// jest.mock('axios')

describe("_App Component test", () => {
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
	beforeEach(() => {
		useDispatchMock.mockClear();
		useDispatchMock.mockReturnValue(jest.fn());

	});
	beforeAll(() => server.listen());

	afterEach(() => server.resetHandlers());

	afterAll(() => server.close());

	test('_ welcome message / app', () => {
		const { getByText } = render(
			<Provider store={store}>
				<App />
			</Provider>
		);

		expect(getByText(/Welcome to DarkTetris/i)).toBeInTheDocument();
	});

	test('_ error message exists / app', () => {
		render(<App />, {
			preloadedState:
			{
				error: { title: "error", message: "error text here..." }
			}
		})
		expect(screen.queryByText("error")).toBeInTheDocument()
		expect(screen.queryByText("error text here...")).toBeInTheDocument()
	});

	test("_ player ingame case / app", () => {
		const costumStore = aallaliStore(
			{
				game: { title: "zribaK7la", tetros: ['L', 'Z', 'O', 'T'] },
				player: { nickname: "l7wli" }
			}
		)
		const tree = renderer.create(<Provider store={costumStore}>
			<App />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});


	// test("_ player exists but not ingmae case / app", () => {
	// 	act(() => {
	// 		render(<App />, {
	// 			preloadedState:
	// 			{
	// 				player: { nickname: "l7wli" }
	// 			}
	// 		})
	// 	})

	// 	expect(screen.queryByText("No open room available ... create yours now!")).toBeInTheDocument()
	// });
	test('_ invalid Hash Query / app', () => {
		window.location.hash = "roomY"
		render(
			<App />)
		expect(screen.queryByText("Invalid Hash Query"))
		expect(window.location.hash).toEqual("#roomY");
	});

	test('_ valid Hash Query / app', () => {
		window.location.hash = "#arena[allali]"
		render(<App />)
		// expect(screen.queryByText("Invalid Hash Query"))
		expect(window.location.hash).toEqual("#arena[allali]");
	});

	// test('_ valid Hash Query with name already set / app', () => {
	// 	window.location.hash = "#arena[allali]"
	// 	render(<App />, {
	// 		preloadedState:
	// 			{ player: { nickname: "karen" } }
	// 	})
	// 	expect(screen.queryByText("karen")).toBeInTheDocument()
	// });


	// test('_ invalid hash query / app', () => {
	// 	// Object.defineProperty(window.location, 'hash', {
	// 	// 	writable: true,
	// 	// 	value: '#arena[allali]'
	// 	// });
	// 	Object.defineProperty(window, 'location', {
	// 		value: {
	// 			hash: "http://localhost/#arena[allali]"
	// 		}
	// 	})

	// 	Object.defineProperty(window, 'hash', {
	// 		value: {
	// 			hash: "#arena[allali]"
	// 		}
	// 	})
	// 	console.log(window.location.hash)
	// 	render(
	// 		<App />)
	// 	console.log("=============>", window.location.hash, window.location.href)

	// 	console.log(screen.debug())

	// 	// expect(screen.queryByText("Invalid Hash Query"))
	// });
})
