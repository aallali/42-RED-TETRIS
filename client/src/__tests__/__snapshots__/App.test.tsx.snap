// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ render app component 1`] = `
<div>
  <div
    className="min-h-screen   flex justify-center items-center"
    style={
      Object {
        "background": "url(\\"tetris_home_wallpaper.jpg\\") no-repeat",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  >
    <div
      className="py-10 px-10 bg-white rounded-2xl shadow-xl z-20 "
      style={
        Object {
          "width": "400px",
        }
      }
    >
      <div>
        <h1
          className="text-3xl font-bold text-center mb-4 cursor-pointer"
        >
          Welcome to DarkTetris
        </h1>
        <p
          className="w-80 text-center text-sm mb-8 font-semibold text-gray-700 tracking-wide cursor-pointer"
        >
          choose a nickname to start playing
        </p>
      </div>
      <div
        className="space-y-4 "
      >
        <input
          className="block text-2xl py-3 px-4 rounded-lg w-full border outline-none"
          onChange={[Function]}
          placeholder="nickname..."
          type="text"
          value=""
        />
        <p
          className="text-red-500 text-md p-1 "
        >
          
        </p>
      </div>
      <div
        className="text-center mt-6"
      >
        <button
          className="py-3 w-64 text-xl font-bold  text-2xl text-white  rounded-2xl bg-gray-900"
          disabled={false}
          onClick={[Function]}
        >
          PLAY
        </button>
      </div>
    </div>
  </div>
</div>
`;
